src/gameState.ts

```diff
@@ -0,0 +1,82 @@
+import { Board } from "./board/board";
+import { loadGameConfigFromFile } from "./gameLoader";
+import { Coordinates, checkIfMoveIsAllowed } from "./move/move";
+import { Player } from "./player";
+import { Tile } from "./tile/tile";
+
+const initialGameState: GameState = {
+  currentPlayer: "black",
+  turnNumber: 1,
+};
+
+export interface GameState {
+  board?: Board;
+  turnNumber: number;
+  currentPlayer: Player;
+  winner?: Player;
+  selectedMoveCordinates?: Coordinates;
+}
+export type GameActionName =
+  | "SELECT_TILE"
+  | "FORFEIT"
+  | "WAITING"
+  | "END_TURN"
+  | "LOAD_BOARD";
+export interface GameStateAction {
+  name: GameActionName;
+  selectedMoveCoordinates?: Coordinates;
+  newBoard?: Board;
+}
+
+const setAllowedTilesOnBoard = (gameState: GameState): Board => {
+  if (gameState.board == null) {
+    return null;
+  }
+
+  if (gameState.turnNumber === 1) {
+    let newBoard: Board = JSON.parse(JSON.stringify(gameState.board));
+    gameState.board.forEach((line, y) =>
+      line.forEach((tile, x) => {
+        const currentTile = newBoard[y][x];
+
+        if (currentTile == null) {
+          return;
+        }
+
+        newBoard[y][x].moveAllowed = checkIfMoveIsAllowed(
+          { x, y },
+          gameState.turnNumber === 1
+        );
+      })
+    );
+    return newBoard;
+  }
+};
+
+export const computeGameState = (
+  gameState: GameState | undefined,
+  action: GameStateAction
+): GameState => {
+  if (action.name === "SELECT_TILE") {
+    const isMoveAllowed = checkIfMoveIsAllowed(
+      gameState.selectedMoveCordinates,
+      gameState.turnNumber === 1
+    );
+
+    if (isMoveAllowed) {
+      return {
+        ...gameState,
+        selectedMoveCordinates: action.selectedMoveCoordinates,
+      };
+    }
+  }
+
+  if (action.name === "LOAD_BOARD") {
+    const board: Board = loadGameConfigFromFile();
+    const newGameState = { ...initialGameState, board: board };
+
+    return { ...newGameState, board: setAllowedTilesOnBoard(newGameState) };
+  }
```

Comment:
I'm a bit surprised to see board initialisation logic here. IMO `computeGameState` should be used to update the game state from a previously (already initialized) game state.

Creating the initial board can be done at program startup, in `index.ts` for example. This is the approach taken by #7  if I'm not mistaken so we should keep it :slightly_smiling_face:

Reply:
Same as https://github.com/marmelab/kamon/pull/7#discussion_r1172496904